lib/Sema/CSDiag.cpp

```diff
@@ -5410,13 +5415,79 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
 
     return true;
   }
-  
+
+
+  // If this is an exact match on arguments or there is only one candidate
+  // available, it might be either contextual conversion or generics error,
+  // let's try to diagnose generics problem here.
+  if (calleeInfo.size() == 1) {
+    auto candidate = calleeInfo[0];
+    // We have an initializer expression for some type (might be generic),
+    // before giving up trying to diagnose this situation, let's double
+    // check if we have any unresolved archetypes, and try to diagnose that.
+    // This might happen in the situation like following:
+    // struct S<A, B> { init(a: A) {} } and call S(a: 1),
+    // initializer is going to be (A) -> S<A, B>, A is inferred from
+    // initializer argument and B going to be unspecified.
+    auto decl = candidate.getDecl();
+    // If the candidate is a valid constructor declaration, let's proceed.
+    if (decl && !decl->isInvalid() && isa<ConstructorDecl>(decl)) {
+      auto resultType = candidate.getResultType();
+      // We can only try to diagnose bound generic types.
+      if (auto BGT = resultType->getAs<BoundGenericType>()) {
+        llvm::SmallSet<ArchetypeType *, 2> missingParams;
+        for (auto &genericParam : BGT->getGenericArgs()) {
+          if (auto archetype = genericParam->getAs<ArchetypeType>()) {
+            // First, let's consider all of the generic arguments missing.
+            missingParams.insert(archetype);
+          }
+        }
+
+        std::function<void(Type)> recordArchtype = [&](Type argumentType) {
+          // Syntax Sugar type e.g. Optional or Array.
+          if (auto SST = dyn_cast<SyntaxSugarType>(argumentType.getPointer()))
+            argumentType = SST->getBaseType();
+
+          // Handles associative collections e.g. Dictionary or Set.
+          if (auto BGT = argumentType->getAs<BoundGenericType>()) {
```

Comment:
This isn't so great—it still only handles one level of nesting. (Consider `Optional<Array<T>>`.) How about using a TypeWalker instead?